### PR TITLE
Port to Collabwrapper 

### PR DESCRIPTION
**Pull request by Saumya-Mishra9129**
Tested , collaboration works fine, both joiner and sharer share same grid, and state is also shared between them , but colors on both grid are different as they take colors from sugar. Do colors should be same?. 
@chimosky @quozl @JuiP @srevinsaju Please review.

### CODE REVIEW

File: DeductoActivity.py [hunk 1]
```diff
@@ -44,21 +37,13 @@
...
     """ Logic puzzle game """
 
     def __init__(self, handle):
         """ Initialize the toolbars and the game board """
-        try:
-            super(DeductoActivity, self).__init__(handle)
-        except dbus.exceptions.DBusException as e:
-            _logger.error(str(e))
+
+        super(DeductoActivity, self).__init__(handle)
```
Comment (srevinsaju) [inline DeductoActivity.py]: Am I right, if `super(DeductoActivity, self).__init__(handle)` can be simplified to `super().__init__(handle)` ?

File: DeductoActivity.py [hunk 2]
```diff
@@ -87,8 +72,13 @@ def __init__(self, handle):
         self._game = Game(canvas, parent=self, colors=self.colors)
 
         self._sharing = False
-        self._initiating = False
-        self._setup_presence_service()
+        self._initiating = None
```
Comment (srevinsaju) [inline DeductoActivity.py]: May I know why this was changed?

**Review by srevinsaju — COMMENTED** [review]
did a code review. Nice! 

I am speculating about `set_data` and `get_data`. Are nothing to be done in those functions?

**Comment by Saumya-Mishra9129** [comment]
402f6b2 is effective but collaboration takes some time, after 1 round of play we can see same colors on both sharer and joiner.

File: DeductoActivity.py (hunk 1 above)
Comment (Saumya-Mishra9129) [reply DeductoActivity.py]: Yeah you are , these arguments were necessary in Python 2 , but in Python 3  we can use either of them. Thanks

File: DeductoActivity.py (hunk 2 above)
Comment (Saumya-Mishra9129) [reply DeductoActivity.py]: Thanks for noticing , I find out that it was a mistake. I will change this one.

**Comment by quozl** [comment]
Keep in mind that https://github.com/sugarlabs/colordeducto is very similar, and it saves effort to cherry-pick patches from one repository to the other.  The repositories have diverged slightly because people haven't kept that in mind.

**Comment by quozl** [comment]
Thanks.

* [x] collabwrapper.py differs from upstream,
* [x] some of your changes don't pass style checks, e.g. space after comma,

Could not test, was unable to find a test case or procedure.  Can you suggest?

**Comment by Saumya-Mishra9129** [comment]
> Thanks.
> 
>  collabwrapper.py differs from upstream,
>  some of your changes don't pass style checks, e.g. space after comma,

Thanks @quozl for review. Fixed in 23863e9.

> Could not test, was unable to find a test case or procedure. Can you suggest?

There are the things I tried while testing collaboration -

- joiner is able to join the sharer.
- Clicking on Start a new game by sharer , starts a new game at joiner. Also only sharer is able to start a new game, joiner cannot.
- Same grid is shared between sharer and joiner ( same colours also).
- If joiner recognises a pattern and clicks on True or False buttons , game state will be updated in sharer and vice-versa.
-  Explore some examples is not shared between sharer and joiner.

**Comment by quozl** [comment]
Thanks.  Thing is, I don't recognise patterns, and there's no instructions.  I haven't had time to read the source code.

**Comment by Saumya-Mishra9129** [comment]
> Thanks. Thing is, I don't recognise patterns, and there's no instructions. I haven't had time to read the source code.

@quozl I have looked at source code. There are 10 levels. Game will start from 1st level , if you give 5 correct answers, you will be updated to next level. These are the patterns for each level. 

- Level 1: Center dot is True
- Level 2: Corners match
- Level 3: more than 12 True
- Level 4: even number of True
- Level 5: diagonal is True
- Level 6: True above each False
- Level 7: True in each row
- Level 8: True path to center
- Level 9: more True on top half than on bottom half
- Level 10: smilely face

Here True is dark coloured dot and False is light coloured one.

**Comment by quozl** [comment]
Thanks.  Tested.

* [ ] you've added a restriction that only the original sharer can press new game button; we don't usually implement this because it places a critical run-time dependency on the sharer's system and her engagement; if they leave and rejoin, nobody will be able to press new game button; we normally let any participant press new game button.
* [ ] the game state is not shared until new game button is pressed, see below.

@srevinsaju asked;
> I am speculating about `set_data` and `get_data`. Are nothing to be done in those functions?

These functions are to share the game state from a sharer to a joiner at the time the joiner joins.  Another method is for the sharer to send the grid when a joiner joins.

**Comment by Saumya-Mishra9129** [comment]
> you've added a restriction that only the original sharer can press new game button; we don't usually implement this because it places a critical run-time dependency on the sharer's system and her engagement; if they leave and rejoin, nobody will be able to press new game button; we normally let any participant press new game button.

Thanks , I haven't added this restriction. It was there already in master branch. But I can change it , remove this restriction.

> the game state is not shared until new game button is pressed, see below.

Thanks, I got it. Now I have to implement set_data and get_data.

**Comment by Saumya-Mishra9129** [comment]
> the game state is not shared until new game button is pressed, see below.

Game state is shared without pressing new game button , but it had some delay. It takes time to update the shared state on receiver side.

**Comment by quozl** [comment]
> Game state is shared without pressing new game button , but it had some delay. It takes time to update the shared state on receiver side.

Why is that?  Can you fix it?  I'm puzzled by the use of two joined signals from different levels of the stack; one for sugar3, and one for CollabWrapper.  It is improper to use CollabWrapper to transmit data before it has sent a joined signal, and the sugar3 joined signal may happen at a different time.

**Comment by Saumya-Mishra9129** [comment]
> Why is that?

I don't have any idea. The code is almost same as colourdeducto. However when I tested colordeducto in https://github.com/sugarlabs/colordeducto/pull/16 , It has less delay as compared to deducto. 

> Can you fix it? Can you fix it? I'm puzzled by the use of two joined signals from different levels of the stack; one for sugar3, and one for CollabWrapper. It is improper to use CollabWrapper to transmit data before it has sent a joined signal, and the sugar3 joined signal may happen at a different time.

I will try not to use two joined signal as you said. But I don't know it will fix this or not. This delay can be seen in many other activities as well.

**Comment by srevinsaju** [comment]
You are right, colordeducto works faster than deducto, but looking at the diff, I am not able to make out why colordeducto works faster. On  colordeducto#16, you have also used two signals, like this one, but it works faster. Maybe colordeducto received another patch which deducto doesn't share?

**Comment by quozl** [comment]
ImageViewer source demonstrates a minimal set of signals.  You might review that source.

To find the cause of the delay, you might add instrumentation (e.g. logging calls) and analyse the logs (sharer and joiner) against a tcpdump taken in parallel.

You might also use Coverage to make sure that you understand what lines are executed.

To resolve a theory that ColorDeducto code has a fix, compare the two source files across the repositories.  I use emacs ediff or meld.  There are some surprising differences in the respective _master_ branches in the `_new_tube_cb`.

**Comment by Saumya-Mishra9129** [comment]
> To resolve a theory that ColorDeducto code has a fix, compare the two source files across the repositories. I use emacs ediff or meld. There are some surprising differences in the respective master branches in the _new_tube_cb.

Thanks tried this but it couldn't help, actually delay is ambiguous. Sometimes I get same game state in just 10 seconds. Sometimes it takes more than 30 seconds.

**Comment by quozl** [comment]
Do you see same delays using Chat activity?

**Comment by Saumya-Mishra9129** [comment]
> Do you see same delays using Chat activity?

Yeah I can see there. 10 seconds of delay and also Online , Offline , Connected timers. I also saw https://github.com/sugarlabs/sugar/issues/820 , You have mentioned here how ping times varies for Ubuntu.

**Comment by quozl** [comment]
Use Linux kernel API tools (ps axww) to determine if any of the Telepathy processes terminate and restart during that test?  These processes are created after D-Bus receives IPC for Telepathy from Sugar, and should persist without PID change until Sugar logs out or restarts.

**Comment by Saumya-Mishra9129** [comment]
> Use Linux kernel API tools (ps axww) to determine if any of the Telepathy processes terminate and restart during that test? These processes are created after D-Bus receives IPC for Telepathy from Sugar, and should persist without PID change until Sugar logs out or restarts.

Thanks I will sure take a look.
Some interesting results can be seen while testing coverage.
**python3-coverage run /usr/bin/sugar-activity3**

```
1597044523.550005 ERROR deducto-activity: org.freedesktop.DBus.Error.NoReply: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
Traceback (most recent call last):
  File "/usr/bin/sugar-activity3", line 5, in <module>
    activityinstance.main()
  File "/usr/lib/python3.7/dist-packages/sugar3/activity/activityinstance.py", line 230, in main
    instance = create_activity_instance(activity_constructor, activity_handle)
  File "/usr/lib/python3.7/dist-packages/sugar3/activity/activityinstance.py", line 59, in create_activity_instance
    activity = constructor(handle)
  File "/usr/share/sugar/activities/deducto/DeductoActivity.py", line 62, in __init__
    self._setup_toolbars()
  File "/usr/share/sugar/activities/deducto/DeductoActivity.py", line 137, in _setup_toolbars
    stop_button = StopButton(self)
  File "/usr/lib/python3.7/dist-packages/sugar3/activity/widgets.py", line 92, in __init__
    activity.add_stop_button(self)
  File "/usr/lib/python3.7/dist-packages/sugar3/activity/activity.py", line 485, in add_stop_button
    self._stop_buttons.append(button)
AttributeError: 'DeductoActivity' object has no attribute '_stop_buttons'

```

It looks like a toolkit issue.

**Comment by chimosky** [comment]
@Saumya-Mishra9129 the error you posted above is expected because on [L46](https://github.com/Saumya-Mishra9129/deducto/blob/major/DeductoActivity.py#L46) `super` is called and the `sugar3.activity.activity` class was never initialised.

**Comment by Saumya-Mishra9129** [comment]
> @Saumya-Mishra9129 the error you posted above is expected because on [L46](https://github.com/Saumya-Mishra9129/deducto/blob/major/DeductoActivity.py#L46) `super` is called and the `sugar3.activity.activity` class was never initialised.

I am sure whether this causes the issue or not. In Python 3  super(DeductoActivity, self).__init__(handle) can be simplified to super().__init__(handle) , both are same. I don't think this causes the error while testing coverage.

**Comment by chimosky** [comment]
@Saumya-Mishra9129 said
> I am sure whether this causes the issue or not. In Python 3 super(DeductoActivity, self).init(handle) can be simplified to super().init(handle) , both are same. I don't think this causes the error while testing coverage.

I agree, it doesn't cause the error and I also don't think it's a toolkit issue.

**Comment by chimosky** [comment]
Tested, previous error doesn't persist although there's a `PyGIWarning` in the logs,
```
PyGIWarning: PangoCairo was imported without specifying a version first. Use gi.require_version('PangoCairo', '1.0') before import to ensure that the right version gets loaded.
  from gi.repository import PangoCairo
```

Collaboration works as expected, although, the joiner's game colours change to the sharer's and is that intended?

**Comment by quozl** [comment]
> AttributeError: 'DeductoActivity' object has no attribute '_stop_buttons'

This error is very clear evidence that the `__init__` has not run.  See https://github.com/sugarlabs/sugar-toolkit-gtk3/blob/68e57eff8458a89704bfc5b22f171514948bacd0/src/sugar3/activity/activity.py#L468

`_stop_buttons` is initialised by Activity's `__init__`, so for it not to have been run is very important to resolve.

_super_ doesn't have the same effect, in my experience, as calling the `__init__` using the class name.

In particular, we use multiple inheritance in Activity, and there are warnings I've found again just now which say _super_ is for single inheritance only.  Also, you cannot use _super_ if the subclasses don't use it, and vice versa.  Since our subclasses don't use _super_, that means we can't use it either.

* [Every occurrence of the method must include super() after you use it](https://www.pythonforbeginners.com/super/working-python-super-function).
* [Subclasses must use super if their superclasses do](https://fuhm.net/super-harmful/)

It would be impractical to change the toolkit to use _super_ because we would have to change all activities at once, and we wouldn't be able to use a new activity safely on an old Sugar, and vice versa.

**Comment by Saumya-Mishra9129** [comment]
> the joiner's game colours change to the sharer's
                                 
I think so , I have implemented collaboration that way here. @quozl and @chimosky Can you look I am not able to see any delay now. Collaboration works fine. Can we merge this PR and open issues for further requirements. ?

**Comment by Saumya-Mishra9129** [comment]
> In particular, we use multiple inheritance in Activity, and there are warnings I've found again just now which say super is for single inheritance only. Also, you cannot use super if the subclasses don't use it, and vice versa. Since our subclasses don't use super, that means we can't use it either.

> Every occurrence of the method must include super() after you use it.
Subclasses must use super if their superclasses do
It would be impractical to change the toolkit to use super because we would have to change all activities at once, and we wouldn't be able to use a new activity safely on an old Sugar, and vice versa.

Thanks new info for me. I have removed super now.

**Comment by quozl** [comment]
Thanks.  Reviewed again.  I've mentioned this before, but I'll put it another way;

* [ ] your code connects to both the Activity _joined_ signal and CollabWrapper's _joined_ signal, specifying a function that sets variables; the same function is connected to the Activity _shared_ signal; it isn't clear why your code does this, as CollabWrapper also connects to the Activity _joined_ signal and cascades to emitting the CollabWrapper _joined_ **or** CollabWrapper may emit the signal independently of the Activity _joined_ signal ... this is confusing; why do you want to connect these signals and is there another way you can do it?

**Comment by Saumya-Mishra9129** [comment]
> your code connects to both the Activity joined signal and CollabWrapper's joined signal, specifying a function that sets variables; the same function is connected to the Activity shared signal; it isn't clear why your code does this, as CollabWrapper also connects to the Activity joined signal and cascades to emitting the CollabWrapper joined or CollabWrapper may emit the signal independently of the Activity joined signal ... this is confusing; why do you want to connect these signals and is there another way you can do it?

Thanks , @quozl Can we not just use collabwrapper joined signal ? As CollabWrapper is used for collaboration, we should just call the joined event of collabwrapper and not the one from the sugar3.activity.activity.Activity class. 
Also I used this way because many activities I have seen implemented collaboration this way.

**Comment by Saumya-Mishra9129** [comment]
Tested till 93ca2e9. Works Fine. Please review.

**Pull request merged**